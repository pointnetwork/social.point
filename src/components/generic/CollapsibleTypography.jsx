import { useEffect, useState } from "react";
import { makeStyles } from '@material-ui/core/styles';
import { Typography } from '@material-ui/core';
import { Skeleton } from '@material-ui/lab';
import processString from 'react-process-string';

const useStyles = makeStyles((theme) => ({
    block: {
        display: 'block',
        cursor:'pointer',
        width: '100%'
    },
    inline: {
        display: 'inline-block',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
        cursor:'pointer',
        width: '100%'
    }
}));

const CollapsibleTypography = ({content, loading}) => {
    const [collapsed, setCollapsed] = useState(true);
    const [rendered, setRendered] = useState(content);
    const styles = useStyles();

    useEffect(() => {

        let config = [{
            regex: /(http|https):\/\/(\S+)\.([a-z]{2,}?)(.*?)( |\,|$|\.)/gim,
            fn: (key, result) => <span key={key}>
                                     <a target="_blank" rel="noreferrer" href={`${result[1]}://${result[2]}.${result[3]}${result[4]}`}>{result[2]}.{result[3]}{result[4]}</a>{result[5]}
                                 </span>
        }, {
            regex: /(\S+)\.([a-z]{2,}?)(.*?)( |\,|$|\.)/gim,
            fn: (key, result) => <span key={key}>
                                     <a target="_blank" rel="noreferrer" href={`http://${result[1]}.${result[2]}${result[3]}`}>{result[1]}.{result[2]}{result[3]}</a>{result[4]}
                                 </span>
        }, {
            regex: /\@([a-z0-9_\-]{1,15}+)( |\,|$|\.)/gim, //regex to match a username
            fn: (key, result) => {
                let username = result[1];
                return <span key={key}><a href={`/profile/${username}`}>@{username}</a>&nbsp;</span>
            }

        }];

        setRendered(processString(config)(content));

    }, []);


    const toggleCollapse = () => {
        setCollapsed(!collapsed);
    }

    return (
        collapsed
        ?
        <Typography component="span" variant="body2" className={styles.inline} color="textPrimary" align="left" noWrap={true} onClick={toggleCollapse}>
            {loading ? <Skeleton /> : rendered }
        </Typography>
        : 
        <Typography component="span" variant="body2" className={styles.block} color="textPrimary" align="left" onClick={toggleCollapse}>
            {loading ? <Skeleton /> : rendered }
        </Typography>            
    );
};

export default CollapsibleTypography