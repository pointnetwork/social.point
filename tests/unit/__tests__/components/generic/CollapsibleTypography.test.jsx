import React from "react";
import renderer from 'react-test-renderer';

import CollapsibleTypography from '../../../../../src/components/generic/CollapsibleTypography';

describe("CollapsibleTypography", () => {

    it('should render', () => {
        const component = renderer.create(
            <CollapsibleTypography content="this is a test" loading={false}/>
        )
        expect(component.toJSON()).toMatchSnapshot();
    });

});