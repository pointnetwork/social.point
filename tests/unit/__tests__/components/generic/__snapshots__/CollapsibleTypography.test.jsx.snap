// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleTypography should render 1`] = `
<span
  className="MuiTypography-root makeStyles-inline-2 MuiTypography-body2 MuiTypography-colorTextPrimary MuiTypography-noWrap MuiTypography-alignLeft"
  onClick={[Function]}
>
  this is a test
</span>
`;
